### DISCUSSION TITLE
cpu limit per isolate worker

### DISCUSSION THREAD

**Discussion by amirhosseinvz94** [denoland/deno#12626] [Q&A]
Hi,

I'm executing untrusted javascript inside deno, and i have trouble
with some scripts that take cpu time.

Is there any way I can set an execution cpu time limit of some deno executed
script?

**Comment by norpol** [comment]
That would be quite awesome indeed. Some more resource requests (so also memory limits) and discussions I could find: #10036,  https://github.com/denoland/deno/discussions/11160, #6916, #12637.

Each worker is running in it's own thread as described in detail in the [internals of deno book](https://choubey.gitbook.io/internals-of-deno/threading-model/default-threads).

You can see this behavior with  `ps` on Linux (I'm starting the official example of deno here.

```bash
$ ps -T -p $(pgrep deno)
    PID    SPID TTY          TIME CMD
  28184   28184 pts/1    00:00:00 deno
  28184   28187 pts/1    00:00:00 V8 DefaultWorke
  28184   28188 pts/1    00:00:00 V8 DefaultWorke
  28184   28189 pts/1    00:00:00 V8 DefaultWorke
  28184   28190 pts/1    00:00:00 worker-0
```

The `SPID` column elements are accessible via `/proc/<pid>/tasks/<spid>`.

The deno worker threads are created via Rust [`std::thread::Builder`](https://doc.rust-lang.org/std/thread/struct.Builder.html)  which in itself allows no specification of any sorts of limits or constraints (See [deno source `runtime/ops/worker_host.rs`](https://github.com/denoland/deno/blob/64e072e499d36ca824db297a493667415ed67cdf/runtime/ops/worker_host.rs#LL196C24-L196C44)).

Implementing resource limits from scratch can be a major challenge as you need to support a variety of platforms:

* Linux with cgroups or at least `nice`ness
* Linux inside Docker / no access to cgroups wouldn't be possible that way
* Windows/Mac ???

In addition you'll have to implement your own separate "IPC rate-limiting", as a child process cold still send an excessive amount of data/messages back to the main deno thread.

An alternative for now would be to simply put duration constrains on workers. This could be accomplished by calling `Worker.terminate()` after a fixed amount of time (like with setTimeout) Could be sufficient in conjunction with some task control.

I couldn't figure this out so quickly, but if you are on Linux and you can live with some half-baked solution it might be possible to look for new `worker-n` threads and then to tell Linux cgroup to limit  those on the fly - not ideal but might be a solution. Not sure if there is a more standard way of setting resource constrains on threads.

Not sure but https://docs.rs/cgroups/latest/cgroups/cgroup_builder/index.html might be good starting point for putting the WebWorkers into each own cgroup per worker thread.

Another suggestion by one deno developer is usng the [`deno_core`](https://crates.io/crates/deno_core)` rust crate: https://github.com/denoland/deno/discussions/11160#discussioncomment-932647

**Reply by ahuigo** [reply]
I found an article: https://deno.com/blog/anatomy-isolate-cloud , and it mentions:

- **Running deployments in separate processes.** 
V8 isolates are already… isolated… from one another. Deno Deploy goes a step further and enlists the operating system’s help to improve the isolation of each deployment.

- **Hypervisor monitoring of resource utilization.**
The runner continuously tracks the metrics of all running deployments. This is necessary for billing purposes, but also allows the runner to enforce resource utilization quotas. Deployments that consume too many resources are terminated to prevent service degradation.
